rust_src/crates/webrender/src/event_loop.rs

```diff
@@ -0,0 +1,275 @@
+
+use futures::future::FutureExt;
+#[cfg(unix)]
+use glutin::platform::unix::EventLoopWindowTargetExtUnix;
+use glutin::{
+    event::{Event, StartCause, WindowEvent},
+    event_loop::{ControlFlow, EventLoop, EventLoopProxy},
+    platform::run_return::EventLoopExtRunReturn,
+    window::{WindowBuilder, WindowId},
+    ContextBuilder, ContextCurrentState, CreationError, NotCurrent, WindowedContext,
+};
+use libc::{c_void, fd_set, sigset_t, timespec};
+use once_cell::sync::Lazy;
+use tokio::{io::unix::AsyncFd, runtime::Runtime, time::Duration};
+
+use crate::future::batch_select;
+
+pub type GUIEvent = Event<'static, i32>;
+
+#[allow(dead_code)]
+#[derive(Clone, Copy)]
+pub enum Platform {
+    X11,
+    Wayland(*mut c_void),
+    MacOS,
+    Windows,
+}
+
+unsafe impl Send for Platform {}
+
+pub struct WrEventLoop {
+    el: EventLoop<i32>,
+}
+
+unsafe impl Send for WrEventLoop {}
+unsafe impl Sync for WrEventLoop {}
+
+impl WrEventLoop {
+    pub fn build_window<'a, T: ContextCurrentState>(
+        &mut self,
+        window_builder: WindowBuilder,
+        context_builder: ContextBuilder<'a, T>,
+    ) -> Result<WindowedContext<NotCurrent>, CreationError> {
+        context_builder.build_windowed(window_builder, &self.el)
+    }
+
+    pub fn create_proxy(&self) -> EventLoopProxy<i32> {
+        self.el.create_proxy()
+    }
+
+    pub fn wait_for_window_resize(&mut self, target_window_id: WindowId) {
+        let deadline = Instant::now() + Duration::from_millis(100);
+        self.el.run_return(|e, _, control_flow| match e {
+            Event::NewEvents(StartCause::Init) => {
+                *control_flow = ControlFlow::WaitUntil(deadline);
+            }
+            Event::NewEvents(StartCause::ResumeTimeReached { .. }) => {
+                *control_flow = ControlFlow::Exit;
+            }
+
+            Event::WindowEvent {
+                event: WindowEvent::Resized(_),
+                window_id,
+            } => {
+                if target_window_id == window_id {
+                    *control_flow = ControlFlow::Exit;
+                }
+            }
+            _ => {}
+        });
+    }
+
+    pub fn detect_platform(&mut self) -> Platform {
+        #[cfg(unix)]
+        {
+            if self.el.is_wayland() {
+                return Platform::Wayland(
+                    self.el
+                        .wayland_display()
+                        .expect("Fetch Wayland display failed"),
+                );
+            } else {
+                return Platform::X11;
+            }
+        }
+        #[cfg(macos)]
+        {
+            return Platform::MacOS;
+        }
+        #[cfg(windows)]
+        {
+            return Platform::Windows;
+        }
+    }
+}
+
+pub static EVENT_LOOP: Lazy<Mutex<WrEventLoop>> = Lazy::new(|| {
+    Mutex::new(WrEventLoop {
+        el: glutin::event_loop::EventLoop::with_user_event(),
+    })
+});
+
+pub static TOKIO_RUNTIME: Lazy<Mutex<Runtime>> =
```

Comment:
One thing I thing looking at this - would there be benefit to having a shared Tokio runtime between JS and the event loop if both js and WebRender are enabled?

Reply:
Using a shared tokio runtime will make emacs-ng spawn less threads.